tests/run/pure_cdef_class_dataclass.py

```diff
@@ -29,3 +29,47 @@ class MyDataclass:
 
     a: int = 1
     self: list = cython.dataclasses.field(default_factory=list, hash=False)  # test that arguments of init don't conflict
+
+
+class DummyObj:
+    def __repr__(self):
+        return "DummyObj()"
+
+
+@cython.dataclasses.dataclass
+@cython.cclass
+class NoInitFields:
+    """
+    >>> NoInitFields()
+    NoInitFields(has_default=DummyObj(), has_factory='From a lambda', neither=None)
+    >>> NoInitFields().has_default is NoInitFields().has_default
+    True
+
+    >>> NoInitFields(1)  # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+    ...
+    TypeError: NoInitFields.__init__() takes 1 positional argument but 2 were given
+    >>> NoInitFields(has_default=1)  # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+    ...
+    TypeError: NoInitFields.__init__() got an unexpected keyword argument 'has_default'
+    >>> NoInitFields(has_factory=1)  # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+    ...
+    TypeError: NoInitFields.__init__() got an unexpected keyword argument 'has_factory'
+    >>> NoInitFields(neither=1)  # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+    ...
+    TypeError: NoInitFields.__init__() got an unexpected keyword argument 'neither'
```

Comment:
I think we can shorten this a bit. It's confusing because the ellipses look like a test separator but are actually right in the middle of each single test.

Also, it's sad that we need a doctest option here, but I guess you added it because what different Python (and PyPy) versions tell you here is difficult to match which simple word exclusions. Maybe it could still help to guard against really unexpected failures if we look at least for the name of the argument in the exception text? That should always appear somewhere, I think.

```suggestion
    >>> NoInitFields(1)  # doctest: +IGNORE_EXCEPTION_DETAIL
    Traceback (most recent call last):
    TypeError: NoInitFields.__init__() takes 1 positional argument but 2 were given

    >>> NoInitFields(has_default=1)  # doctest: +ELLIPSIS
    Traceback (most recent call last):
    TypeError: ...has_default...
    >>> NoInitFields(has_factory=1)  # doctest: +ELLIPSIS
    Traceback (most recent call last):
    TypeError: ...has_factory...
    >>> NoInitFields(neither=1)  # doctest: +ELLIPSIS
    Traceback (most recent call last):
    TypeError: ...neither...
```

Reply:
> Also, it's sad that we need a doctest option here, but I guess you added it because what different Python (and PyPy) versions tell you here is difficult to match which simple word exclusions. 

It was actually pure Vs Cython on the same version of Python. I think because the standard library implementation monkey-patches the methods into the class you end up with differences like `__init__` instead of `NoInitFields.__init__`, and differences in argument counts.